### PR TITLE
[feat]: prepare FSDP to handle multiple flatten params and fixed metadata saving for MoE

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?

This is step 3 to address #697. In this PR, we update FSDP to deal with flatten and nonflatten params in a more uniform fashion. The API is unchanged and still only allow either all flatten or none flatten. But internally, FSDP keeps flatten and nonflatten params in `m.params` in an uniform way.

There is no more `m.module` pointing to a FPW or not. It is always a FPW regardless whether flatten is used. When none of the parameter is flattened, FPW is just a noop wrapper, it simply won't have any FlatParameter objects in its `self.flat_params` list.

After this change, FSDP won't care there is 0 or more flat params and it delegates most flatten related logic to FPW. It retains the sharding related logic as well as logic related to autograd hooks etc.

It was tricky to update `local_metadata_dict` and `consolidate_shard_weights` APIs for this change, but the resulting code is about half as long as before. 

I also added a feature to allow users to name their flat parameters in the future. This will be needed when they want to name different groups of params and later use them in the optimizer.

In the next step, I will add API changes to allow users to group params for flatten for real.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -293,18 +307,31 @@ def __init__(
...
-            self.params = params
+            to_be_flatten_params = [params]
+            non_flatten_params = []
+            param_name_groups = [param_names]
+        del param_names
+
+        self._fsdp_wrapped_module: nn.Module = FlattenParamsWrapper(module, param_list=to_be_flatten_params)
+        del module  # free original module in case it helps garbage collection
+
+        # Now, in this FSDP wrapper class, we keep a list of to-be-flatten and remining params
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:325]: (typo) remining

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1529,58 +1560,36 @@ def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
...
-            # used in the FlattenParamsWrapper
-            else:
-                param_names = []
-                for module_path, param_name in m.param_path_infos:
-                    full_param_path = module_path + "." + param_name if module_path else param_name
-                    param_names.append(_clean_path(full_param_path))
-                params_metadata.append(
-                    {
-                        "fsdp_path": _clean_path(path),
-                        "is_flat": True,
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: Is it an issue that we don't retain the 
`no_broadcast_optim_state` and `is_flat` keys?

I am about to integrate `local_metadata_dict`/`consolidate_shard_weights` into `fairseq:gshard` which will get merged into `fairseq:master` soon, so it would be good if there was test coverage for `MixtureOfExperts` in `test_fsdp_metadata.py`.

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: The is_flat entry is no longer needed. The no_broadcast flag was not used by the consolidation function.

Can you share a test case for MoE? I can add it to the unit test and make sure it is covered. For this PR, I made sure the metadata and consolidation functions are consistently updated and unit test was passing. But it is possible moe case is somewhat special?

Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: Its broken before your change :( 
https://github.com/facebookresearch/fairscale/pull/747
we definitely need `no_broadcast_optim_state`. We can probably do without `is_flat`.

Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: It's also broken for vanilla transformers.

I'm looking at it today, but won't be offended if this gets merged while I'm debugging :)

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: Thanks for the PR with supplied tests! I am not in a hurry to merge. I will try to test my new changes with your tests. If making the current master working with your test is a lot of work, there is no need for you to spend too much time on it since I am changing this code anyway. It is better for me to make the new code work with the supplied test cases.

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: > 
> 
> It's also broken for vanilla transformers because it doesn't consider `shared_param_info`.
> 
> I'm looking at it today, but won't be offended if this gets merged while I'm debugging :)

That's a great point on shared_param_info. It is now part of the FlatParameter class and FPW proxy the information to the caller. It should be similar to how I handle the param_info data structure. You can directly work on my branch here if you prefer that. I do think my current code is cleaner than the master version. :-)

Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: Lol I read this after I fixed :(

Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: pushed fixes here!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -1529,58 +1560,45 @@ def _free_full_params(self, params: Optional[List[Parameter]] = None) -> None:
...
+                        backing_param_name = m._param_name_groups[i][0]
+                        names = [backing_param_name]
+                        shapes = [p._orig_size]
+                        numels = [p._orig_size.numel()]
+                    backing_param_name = _clean_path(backing_param_name)
+                    metadata["params"][backing_param_name] = {
+                        "names": [_clean_path(n) for n in names],
+                        "shapes": shapes,
+                        "numels": numels,
+                        "padding_of_backing": m.numel_padded_per_param[i],
```
Comment (sshleifer) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1596]: I would just call this padding

the of_backing names make sense internally but not to a random caller reading a returned dictionary.

**Review by sshleifer — APPROVED** [review]
LGTM, and now I've played with it!  The sooner you merge the better for me :)

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1570]: That's awesome, Sam! Really appreciate the team work here!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -293,18 +307,31 @@ def __init__(
...
-        else:
-            self._fsdp_wrapped_module = module
-            self.param_paths = param_names
-            self.params = params
+            to_be_flatten_params = [params]
+            non_flatten_params = []
+            param_name_groups = [param_names]
+        del param_names
+
+        self._fsdp_wrapped_module: nn.Module = FlattenParamsWrapper(module, param_list=to_be_flatten_params)
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:322]: +1 on this change and making it consistent.

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 5]
```diff
@@ -115,19 +127,27 @@ class FlattenParamsWrapper(nn.Module):
...
+            The module to wrap.
         param_list (Optional[List[List[nn.Parameter]]]):
-            only flatten parameters appearing in the given groups
+            Only flatten parameters appearing in the given groups.
+            If the param_list is an empty list, then no parameters will get flattened.
+            Note, if a single param is in one of the list, it still get flattened and the
+            original param is removed and replaced with the flatten one.
             Default: None, flatten all parameters (if any)
+        flat_param_names (Optional[List[str]]):
+            orignally, give each flat_param a unique name. Note a "flat_param_"
```
Comment (anj-s) [inline fairscale/nn/misc/flatten_params_wrapper.py:146]: type(orignally)

File: tests/nn/data_parallel/test_fsdp_metadata.py [hunk 6]
```diff
@@ -2,14 +2,23 @@
...
 import torch.distributed as dist
 import torch.multiprocessing as mp
 import torch.nn as nn
+from torch.optim import Adam
 
 from fairscale.nn import FullyShardedDataParallel
 from fairscale.utils.testing import in_temporary_directory, skip_if_single_gpu, temp_files_ctx
+from tests.nn.data_parallel.test_fsdp import DistributedTest, MixtureOfExperts, rename_test, spawn_and_init
+
+USE_TEMPFILE = True  # False for debugging
```
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_metadata.py:21]: Should we remove this?

**Review by anj-s — APPROVED** [review]

**Comment by min-xu-ai** [comment]
Once the CI passes, I will merge this in.

@anj-s, we can make a release today for fairseq after this in. I can make it if you are not already planning to do it. Just let me know. Thanks!

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 7]
```diff
@@ -220,41 +241,50 @@ def flat_param(self) -> nn.Parameter:
...
                     else:
-                        shared_param_memo[p] = (m, n)
+                        shared_param_memo[p] = (module_name, m, n)
                         param_infos.append((module_name, m, n))
                         params.append(p)
         del shared_param_memo
 
-        assert len(set(p.dtype for p in params)) == 1, "expects all parameters in module to have same dtype"
+        assert len(set(p.dtype for p in params)) == 1, "expects all parameters to have same dtype"
+        assert len(set(p.requires_grad for p in params)) == 1, "expects all parameters to have same requires_grad"
```
Comment (zhaojuanmao) [inline fairscale/nn/misc/flatten_params_wrapper.py:277]: are you planning to skip flattening parameters if requires_grad are not the same in params?

Comment (min-xu-ai) [reply fairscale/nn/misc/flatten_params_wrapper.py:277]: no, I am going to assume in the future user will just group their params in that use case. I.e. group the params with requires_grad==True together and requires_grad==False together. Then, this assert won't trigger since those will be different FlatParameters.
